[PP-1358] add new events

dbernstein opened the pull request:
## Description
This update captures five new circulation events:

    circulation_manager_hold_expired
    circulation_manager_hold_ready
    circulation_manager_hold_converted_to_loan
    circulation_manager_loan_expired
    circulation_manager_loan_converted_to_hold

I have a couple of questions/comments to inform the review:

1. We are capturing hold expirations  where the CM is and is not the source of truth.  However for loans I believe we are only handling expired loans where the CM is not source of truth. Did I miss something here? Or should we be doing something for loan expirations where CM is not the source of truth?

2. When we convert holds to loans I am both capturing the conversion event as well as recording a "checkout".   Because it is a checkout event that happens to be a conversion.  I think that's what we want but others may disagree. 
 
3. I put a note in the code, but I'll call it out again here:  there appears to be a case where a loan can be converted to a hold.  I'm not sure what the scenario is, but I've captured it is a distinct event in addition to the placing of the hold.  

4. I ran into some mypy issues with the collect_event method. That led me to observe that all loans and holds have patrons and the intention seems to be that all patrons must have loans and holds but the database does not enforce that constraint.  I included the migration here, but I am happy to move it to a separate PR is that seems cleaner.
## Motivation and Context
https://ebce-lyrasis.atlassian.net/browse/PP-1804
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
New unit tests added.  Older unit tests amended.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

dbernstein commented on alembic/versions/20241119_58b0ae7f5b67_make_loan_patron_id_and_hold_patron_id_.py:
```diff
@@ -0,0 +1,43 @@
+depends_on = None
+
+
+def upgrade() -> None:
```
I looked at California.

jonathangreen commented on src/palace/manager/sqlalchemy/model/patron.py:
```diff
@@ -64,12 +64,8 @@ def work(self) -> Work | None:
-        # If this Loan/Hold belongs to an external patron, there may be no library.
-        return None
+    def library(self) -> Library:
+        return self.patron.library
```
I'm going to resolve this one. I did some more looking into type hints for our SQLAlchemy stuff and I was wrong. I think your change is the way to go here.

dbernstein commented on src/palace/manager/celery/tasks/opds_odl.py:
```diff
@@ -123,7 +153,9 @@ def remove_expired_holds(task: Task) -> None:
     for collection_id, collection_name in collections:
         with task.transaction() as session:
-            removed = remove_expired_holds_for_collection(session, collection_id)
+            removed = remove_expired_holds_for_collection(
```
I broke the holds by collection processing out into a separate task to increase the throughput.   I don't think things are likely to take more than 30-45 seconds now, but I can imagine in the future having a substantially larger volume of holds and/or the possibility of the system being down for some significant period of time and needing to catch up.    I think parallelizing the tasks makes a lot of sense here.

dbernstein commented on src/palace/manager/celery/tasks/opds_odl.py:
```diff
@@ -96,6 +119,12 @@ def recalculate_holds_for_licensepool(
+                license_pool=hold.license_pool,
+                event_type=CirculationEvent.CM_HOLD_READY_FOR_CHECKOUT,
+                patron=hold.patron,
+            )
```
Great idea: done.

dbernstein commented on src/palace/manager/core/monitor.py:
```diff
@@ -897,6 +904,9 @@ def delete(self, row):
         """
         self._db.delete(row)
 
+    def post_delete(self, row) -> None:
```
I changed this a bit, but kept the basic idea here which is to ensure that we can log the events after the delete operations have been committed.  See api/monitor.py to see the overriding of `after_commit`

dbernstein commented on src/palace/manager/core/monitor.py:
```diff
@@ -880,11 +880,18 @@ def run_once(self, *args, **kwargs):
         count = qu.count()
         self.log.info("Deleting %d row(s)", count)
         while count > 0:
+            post_delete_ops = []
```
removed.

dbernstein commented on src/palace/manager/celery/tasks/opds_odl.py:
```diff
@@ -7,19 +7,41 @@
-def remove_expired_holds_for_collection(db: Session, collection_id: int) -> int:
+def remove_expired_holds_for_collection(
+    db: Session, collection_id: int, analytics: Analytics
+) -> int:
```
See my comments below:  for this and other reasons you raised, I broke the expired hold removal by collection into its own task.

jonathangreen commented:
@dbernstein looks like mypy is failing on this one

dbernstein commented:
@jonathangreen : this one should be sorted now.

dbernstein commented:
This one has is ready to go.  I fixed the mypy issues a couple of bugs, and reset the migration target.

dbernstein commented:
@jonathangreen ^^

jonathangreen reviewed (commented):
A couple more things to look at here @dbernstein. I'd like to get the race condition when deleting holds sorted out, and make sure we're not going to run into issues referencing a closed session before this gets merged.

dbernstein commented:
@jonathangreen : Hopefully I've addressed your concerns here.  I wanted to keep the event collection outside the transaction as you initially suggested to limit the duration of any row level locks in the database.  To accomplish this requirement I am starting a new session and reattaching the objects from the closed session.  I also removed the migration.

jonathangreen reviewed (approved):
Thanks for working through the changes on this one. Looks good to me!
